3-enrich/scala-common-enrich/src/main/scala/com.snowplowanalytics.snowplow.enrich/common/enrichments/registry/CampaignAttributionEnrichment.scala

```diff
@@ -132,35 +143,23 @@ case class CampaignAttributionEnrichment(
   /**
    * Extract the marketing fields from a URL.
    *
-   * @param uri The URI to extract
+   * @param nvPairs The querystring to extract
    *        marketing fields from
-   * @param encoding The encoding of
-   *        the URI being parsed
    * @return the MarketingCampaign
    *         or an error message,
    *         boxed in a Scalaz
    *         Validation
    */
-  def extractMarketingFields(uri: URI, encoding: String): ValidationNel[String, MarketingCampaign] = {
-
-    val parameters = try {
-      URLEncodedUtils.parse(uri, encoding)
-    } catch {
-      case _ => return "Could not parse uri [%s]".format(uri).failNel[MarketingCampaign]
-    }
-
-    // Querystring map
-    val sourceMap: SourceMap = parameters.map(p => (p.getName -> p.getValue)).toList.toMap
+  def extractMarketingFields(nvPairs: SourceMap): ValidationNel[String, MarketingCampaign] = {
+    val medium = getFirstParameter(mktMedium, nvPairs)
+    val source = getFirstParameter(mktSource, nvPairs)
+    val term = getFirstParameter(mktTerm, nvPairs)
+    val content = getFirstParameter(mktContent, nvPairs)
+    val campaign = getFirstParameter(mktCampaign, nvPairs)
 
-    val decodeString: TransformFunc = CU.decodeString(encoding, _, _)
+    val (clickId, network) = Unzip[Option].unzip(mktClickId.find(pair => nvPairs.contains(pair._1)).map(pair => (nvPairs(pair._1), pair._2)))
```

Comment:
This line will be a bit easier to grok when its `clickIdParameters.find` rather than `mktClickId.find`